File: kBUDGETApp/Sources/Shared/Services/BudgetCalculator.swift
import Foundation

struct BudgetSnapshot {
    let categoryId: UUID
    let allocated: Double
    let carriedOver: Double
    let spent: Double

    var remaining: Double {
        max(0, (allocated + carriedOver) - spent)
    }
}

struct TripProgress {
    let totalSaved: Double
    let target: Double
    let delta: Double
    let onTrack: Bool
}

protocol BudgetCalculating {
    func snapshot(for category: Category, transactions: [Transaction], period: DateInterval) -> BudgetSnapshot
    func tripProgress(goal: TripGoal, transactions: [Transaction], upTo date: Date) -> TripProgress
    func dailyAllowance(remainingThisMonth: Double, daysRemaining: Int) -> Double
    func projectedMonthEnd(spendLast7Days: Double, daysObserved: Int, daysRemaining: Int) -> Double
}

struct BudgetCalculator: BudgetCalculating {
    private let calendar: Calendar

    init(calendar: Calendar = .current) {
        self.calendar = calendar
    }

    func snapshot(for category: Category, transactions: [Transaction], period: DateInterval) -> BudgetSnapshot {
        let spent = transactions
            .filter { $0.category.id == category.id && period.contains($0.date) }
            .reduce(0) { $0 + $1.amountCAD }

        let allocated = category.monthlyLimitCAD
        return BudgetSnapshot(categoryId: category.id, allocated: allocated, carriedOver: 0, spent: spent)
    }

    func tripProgress(goal: TripGoal, transactions: [Transaction], upTo date: Date) -> TripProgress {
        let savingsTransactions = transactions
            .filter { $0.category.name == "Trip Fund" && $0.date <= date }
            .reduce(0) { $0 + $1.amountCAD }

        let totalSaved = goal.currentSavedCAD + savingsTransactions
        let delta = totalSaved - goal.targetCAD
        return TripProgress(totalSaved: totalSaved, target: goal.targetCAD, delta: delta, onTrack: delta >= 0)
    }

    func dailyAllowance(remainingThisMonth: Double, daysRemaining: Int) -> Double {
        guard daysRemaining > 0 else { return 0 }
        return remainingThisMonth / Double(daysRemaining)
    }

    func projectedMonthEnd(spendLast7Days: Double, daysObserved: Int, daysRemaining: Int) -> Double {
        guard daysObserved > 0 else { return 0 }
        let average = spendLast7Days / Double(daysObserved)
        return average * Double(daysRemaining)
    }
}

File: kBUDGETReactNative/src/services/budgetCalculator.ts
import { Category, Transaction, TripGoal } from '@/models';

export interface BudgetSnapshot {
  categoryId: string;
  allocated: number;
  carriedOver: number;
  spent: number;
  remaining: number;
}

export interface TripProgress {
  totalSaved: number;
  target: number;
  delta: number;
  onTrack: boolean;
}

export const snapshotForCategory = (
  category: Category,
  transactions: Transaction[],
  monthStart: Date,
  monthEnd: Date
): BudgetSnapshot => {
  const spent = transactions
    .filter((t) => t.categoryId === category.id)
    .filter((t) => {
      const date = new Date(t.date);
      return date >= monthStart && date <= monthEnd;
    })
    .reduce((sum, t) => sum + t.amountCAD, 0);

  const remaining = Math.max(0, category.monthlyLimitCAD + 0 - spent);

  return {
    categoryId: category.id,
    allocated: category.monthlyLimitCAD,
    carriedOver: 0,
    spent,
    remaining
  };
};

export const tripProgress = (goal: TripGoal, transactions: Transaction[], upTo: Date): TripProgress => {
  const saved = transactions
    .filter((t) => t.categoryId === 'trip-fund')
    .filter((t) => new Date(t.date) <= upTo)
    .reduce((sum, t) => sum + t.amountCAD, 0);

  const totalSaved = goal.currentSavedCAD + saved;
  const delta = totalSaved - goal.targetCAD;
  return {
    totalSaved,
    target: goal.targetCAD,
    delta,
    onTrack: delta >= 0
  };
};

export const dailyAllowance = (remaining: number, daysRemaining: number): number => {
  if (daysRemaining <= 0) {
    return 0;
  }
  return remaining / daysRemaining;
};

export const projectedMonthEnd = (
  spendLast7Days: number,
  daysObserved: number,
  daysRemaining: number
): number => {
  if (daysObserved <= 0) {
    return 0;
  }
  const average = spendLast7Days / daysObserved;
  return average * daysRemaining;
};
